### PR TITLE
Travis-CI testing

### PR DESCRIPTION
**Pull request by anders-dc**
This PR implements Travis CI for CICE. The configuration is similar to Icepack, and is based on GCC and open-mpi.

There are still a few issues that need to be worked out before I recommend merging this PR. The only tests that currently succeed are the `build` tests––the `run` tests all fail. [Here is an example build log](https://api.travis-ci.org/v3/job/359471500/log.txt), with an excerpt below:

```
#------- 
#repo = https://github.com/anders-dc/CICE.git
#bran = 
#hash = 960aaadcf40762e984dd7a75ea36b96df8feef8b
#hshs = 960aaadcf4
#hshu = Anders Damsgaard <andersd@riseup.net>
#hshd = Wed Mar 28 13:20:19 2018 -0400
#date = 2018-03-28
#time = 17:24:42
#mach = travisCI
#user = travis
#vers = CICE 6.0.0.alpha
#------- 
#---
PASS travisCI_gnu_smoke_gx3_8x2_diag1_run5day build
FAIL travisCI_gnu_smoke_gx3_8x2_diag1_run5day run
#---
PASS travisCI_gnu_smoke_gx3_8x2_diag24_medium_run1year build
FAIL travisCI_gnu_smoke_gx3_8x2_diag24_medium_run1year run
#---
PASS travisCI_gnu_smoke_gx3_4x1_debug_diag1_run5day build
FAIL travisCI_gnu_smoke_gx3_4x1_debug_diag1_run5day run
#---
PASS travisCI_gnu_smoke_gx3_8x2_debug_diag1_run5day build
FAIL travisCI_gnu_smoke_gx3_8x2_debug_diag1_run5day run
#---
PASS travisCI_gnu_smoke_gx3_4x2_diag1_run5day build
FAIL travisCI_gnu_smoke_gx3_4x2_diag1_run5day run
#---
PASS travisCI_gnu_smoke_gx3_4x1_diag1_run5day_thread build
FAIL travisCI_gnu_smoke_gx3_4x1_diag1_run5day_thread run
#---
PASS travisCI_gnu_restart_gx3_8x1_diag1 build
PEND travisCI_gnu_restart_gx3_8x1_diag1 exact-restart
FAIL travisCI_gnu_restart_gx3_8x1_diag1 run-initial
#---
PASS travisCI_gnu_restart_gx3_4x2_debug build
PEND travisCI_gnu_restart_gx3_4x2_debug exact-restart
FAIL travisCI_gnu_restart_gx3_4x2_debug run-initial
#---
PASS travisCI_gnu_restart_gx3_8x2_diag1_pondcesm build
PEND travisCI_gnu_restart_gx3_8x2_diag1_pondcesm exact-restart
FAIL travisCI_gnu_restart_gx3_8x2_diag1_pondcesm run-initial
#---
PASS travisCI_gnu_restart_gx3_8x2_diag1_pondtopo build
PEND travisCI_gnu_restart_gx3_8x2_diag1_pondtopo exact-restart
FAIL travisCI_gnu_restart_gx3_8x2_diag1_pondtopo run-initial
#---
PASS travisCI_gnu_smoke_gx1_32x1_diag1_run5day_thread build
FAIL travisCI_gnu_smoke_gx1_32x1_diag1_run5day_thread run
#---
PASS travisCI_gnu_smoke_gx1_16x2_diag1_run5day build
FAIL travisCI_gnu_smoke_gx1_16x2_diag1_run5day run
#---
PASS travisCI_gnu_smoke_gx1_8x4_debug_run2day build
FAIL travisCI_gnu_smoke_gx1_8x4_debug_run2day run
#---
PASS travisCI_gnu_restart_gx1_32x1 build
PEND travisCI_gnu_restart_gx1_32x1 exact-restart
FAIL travisCI_gnu_restart_gx1_32x1 run-initial
#---
PASS travisCI_gnu_restart_gx1_13x2 build
PEND travisCI_gnu_restart_gx1_13x2 exact-restart
FAIL travisCI_gnu_restart_gx1_13x2 run-initial

15 of 36 tests PASSED
15 of 36 tests FAILED
6 of 36 tests PENDING
```

I set `ICE_MACHINE_TPNODE = 4` in [configuration/scripts/machines/env.travisCI](), which makes the build steps succeed. However, Travis-CI does not support the resultant `nprocs` values during execution. By `grep`'ing the generated casescripts, `nprocs` ends up with values of 4, 8, 13, 16, or 32. This, by far, exceeds the capabilities of Travis. **I suggest designing tests that are suitable for Travis**.

Furthermore, I had to remove `-Wextra` from the compiler flags ([configuration/scripts/machines/Macros.travisCI]()), as Travis fails a build if the size of STDOUT/STDERR text exceeds 4 megabytes.

---

*Developer(s):* Anders Damsgaard, Princeton/NOAA-GFDL ([github.com/anders-dc](https://github.com/anders-dc), [adamsgaard.dk](https://adamsgaard.dk))

*Are the code changes bit for bit, different at roundoff level, or more substantial?* There are minor changes to the underlying code which shouldn't affect other uses.

*Is the documentation being updated with this PR? (Y/N)* No.

*If not, does the documentation need to be updated separately? (Y/N)* No.

### CODE REVIEW

**Comment by apcraig** [comment]
I see this in the build log for several tests
 (abort_ice)ABORTED: 
 (abort_ice) error = ice: Input nprocs not same as system request
That suggests there is an inconsistency in the tasks/threads used for testing and those defined by the test and/or in namelist.  What we need to do is make sure we're setting up tests that can be carried out by travisCI.

Does travisCI support MPI and/or openMP and if so, how many tasks and threads can we have?

**Comment by anders-dc** [comment]
Whoops, I forgot to launch the tests with `mpirun`. I've fixed that in a143396 and 
ed04056. TravisCI does support MPI and OpenMP, but the virtual machines are two cores only. However, I think the question is if we can overload the system with additional threads, which would presumably result in slower execution.

**Comment by apcraig** [comment]
It looks like the travisCI machine setup has 4 tasks per node, and also no way to request resources, we are just running interactively.  How many resources do we get, just one node?  That means we may have to develop a suite that uses no more than 4 tasks*threads for all tests.

**Comment by anders-dc** [comment]
Yes, I set it up for 4 tasks per node in order to be able to build. We get just one node with two cores, and I'm pretty sure these are not hyperthreaded. I agree, the best solution would be to have a test suite which is designed for this environment.

**Comment by apcraig** [comment]
We can do that.  We'll need to setup a suite of test that use less resources than we have currently defined.  It would be nice to get access to more cores though so we can test a mix of task and thread counts with different decompositions.  8 or 16 would be great for instance.  

I was just looking to see what VIC is doing and it looks like they use travis for a bunch of unit tests, https://travis-ci.org/UW-Hydro/VIC, but I will try to ask them about whether they are able to test on higher pe counts.

**Comment by anders-dc** [comment]
Sounds good, thanks!

Meanwhile, it looks like we are getting there. I encounter into this error:
```
Fortran runtime error: Cannot open file '/home/travis/CICE_data/grid/gx3/grid_gx3.bin': No such file or directory
```
We used a `wget` call to get external [Icepack_data.tar.gz from a UCAR FTP server]. Is there a similar archive for CICE?

**Comment by apcraig** [comment]
We need to create a small input dataset for travis testing containing just a small gx3 dataset and a set of tests focused on travis resources and the gx3 grid.  I will try to do that now on cheyenne so Anders might have something that works.

**Comment by anders-dc** [comment]
Great! I now fetch the [CICE_data_ic_grid.tar.gz data file](ftp://ftp.cgd.ucar.edu/archive/Model-Data/CICE/CICE_data_ic_grid.tar.gz) before running the tests. However, some of the runs exceed Travis' default tolerance for runtime, so I think it makes a lot of sense to write smaller tests with reduced core and dataset requirements.

**Comment by apcraig** [comment]
@anders-dc if you pull my "travis" branch from my github fork, https://github.com/apcraig/CICE/tree/travis into your branch and activate the "travis_suite" instead of the base_suite, lets see what happens.  I am still working on creating a smaller input dataset.  For now, just download the big file if it's not too time consuming.

**Comment by apcraig** [comment]
With regard to input datasets, there are two tar files that cover what's needed,

ftp://ftp.cgd.ucar.edu/archive/Model-Data/CICE/CICE_data_ic_grid.tar.gz
ftp://ftp.cgd.ucar.edu/archive/Model-Data/CICE/CICE_data_forcing_gx3_all.tar.gz

The first is 28Mb, the second 428 Mb (tarred and zipped).  We could reduce the first a bit as it contains the gx1 grids.  See if that amount of data is allowed and whether that works and helps.

**Comment by apcraig** [comment]
And if we create a gx3 specific tar file with grids, ic, and forcing, that will be 650Mb unzipped.  The full dataset including gx1 forcing was over 20 Gb.

**Comment by anders-dc** [comment]
Thanks Tony. There is no documentation on disk space limits. I tried downloading `CICE_data_all.tar.gz` and the `wget` command made it halfway through. From that crude assessment, there should be approx 10 Gb of usable disk space.

**Comment by apcraig** [comment]
That seems reasonable.  We should be able to do gx3 testing with less than 1 GB.  thanks @anders-dc for your efforts!

**Comment by anders-dc** [comment]
Excellent, thank you Tony. The new test suite seems to mostly succeed.
```
#------- 
#repo = https://github.com/anders-dc/CICE.git
#bran = 
#hash = 76c37bf34ea295fbd2ad889375696104e6e50c7e
#hshs = 76c37bf34e
#hshu = Anders Damsgaard <andersd@riseup.net>
#hshd = Wed Mar 28 18:04:22 2018 -0400
#date = 2018-03-28
#time = 22:08:40
#mach = travisCI
#user = travis
#vers = CICE 6.0.0.alpha
#------- 
#---
PASS travisCI_gnu_smoke_gx3_1x2_diag1_run5day build
PASS travisCI_gnu_smoke_gx3_1x2_diag1_run5day run
#---
PASS travisCI_gnu_smoke_gx3_2x1_debug_diag1_run5day build
FAIL travisCI_gnu_smoke_gx3_2x1_debug_diag1_run5day run
#---
PASS travisCI_gnu_smoke_gx3_1x2_debug_diag1_run5day build
FAIL travisCI_gnu_smoke_gx3_1x2_debug_diag1_run5day run
#---
PASS travisCI_gnu_smoke_gx3_1x1_diag1_run5day_thread build
FAIL travisCI_gnu_smoke_gx3_1x1_diag1_run5day_thread run
#---
PASS travisCI_gnu_smoke_gx3_2x1_diag1_run5day_thread build
PASS travisCI_gnu_smoke_gx3_2x1_diag1_run5day_thread run
FAIL travisCI_gnu_smoke_gx3_2x1_diag1_run5day_thread bfbcomp travisCI_gnu_smoke_gx3_1x2_diag1_run5day.travisCItest different-data
#---
PASS travisCI_gnu_restart_gx3_2x1_diag1 build
PASS travisCI_gnu_restart_gx3_2x1_diag1 run-initial
PASS travisCI_gnu_restart_gx3_2x1_diag1 run-restart
PASS travisCI_gnu_restart_gx3_2x1_diag1 exact-restart
#---
PASS travisCI_gnu_restart_gx3_1x2_diag1 build
PASS travisCI_gnu_restart_gx3_1x2_diag1 run-initial
PASS travisCI_gnu_restart_gx3_1x2_diag1 run-restart
PASS travisCI_gnu_restart_gx3_1x2_diag1 exact-restart
#---
PASS travisCI_gnu_restart_gx3_2x1_diag1_pondcesm build
PASS travisCI_gnu_restart_gx3_2x1_diag1_pondcesm run-initial
PASS travisCI_gnu_restart_gx3_2x1_diag1_pondcesm run-restart
PASS travisCI_gnu_restart_gx3_2x1_diag1_pondcesm exact-restart
#---
PASS travisCI_gnu_restart_gx3_2x1_diag1_pondtopo build
PASS travisCI_gnu_restart_gx3_2x1_diag1_pondtopo run-initial
PASS travisCI_gnu_restart_gx3_2x1_diag1_pondtopo run-restart
PASS travisCI_gnu_restart_gx3_2x1_diag1_pondtopo exact-restart

23 of 27 tests PASSED
4 of 27 tests FAILED
0 of 27 tests PENDING
```
Travis decides to terminate it as it loops through the runlogs in `after_failure` because of the excessive output.

**Comment by apcraig** [comment]
Can we try again, but instead of writing the entire log file at the end, can we just tail -100 each log file?

**Comment by anders-dc** [comment]
Great, this is more informative. Here's the [raw log](https://api.travis-ci.org/v3/job/359611521/log.txt).

**Comment by apcraig** [comment]
We've seen that error before, probably just need to fix an interface call on the cice side.  there is another different error for the 1x1 case.  i'll have to look at that one a little closer.

**Comment by apcraig** [comment]
Just FYI that I have duplicated these errors on another machine with the gnu compiler and am working on them. Hope to have an update soon.

**Comment by apcraig** [comment]
@anders-dc I just updated my travis branch again with several fixes.
https://github.com/apcraig/CICE/tree/travis
The specific commit is
https://github.com/apcraig/CICE/commit/ad52ab44ef77a9d039c108a225ccd292f84186f9
I assume you can pull these updates into your branch and run another test?  If you have problems with the pull, let me know.  thanks!

**Comment by anders-dc** [comment]
Thanks @apcraig, [here's the newest run](https://travis-ci.org/anders-dc/CICE/builds/360360292).

**Comment by apcraig** [comment]
I'm watching it.  We've already hit the log size limit and been going 20 minutes.  We need to add an option to the scripts that doesn't write build output to the terminal.  I will take care of that next.  I've also had an idea that we should be reusing binaries if we can.  That's not so easy to do with CICE because the decomposition is built into the build.  But maybe we can for travis.  Let me prototype that too and see if I can get something that works.

**Comment by apcraig** [comment]
It failed, but we can't tell why.  I'll try to fix the length of the logging and propose another pull later today.

**Comment by anders-dc** [comment]
I agree although the [raw log](https://api.travis-ci.org/v3/job/360360293/log.txt) is still going. The `-Wzerotrip` compiler warnings ("Warning: DO loop at (1) will be executed zero times") is also a major issue.

**Comment by apcraig** [comment]
You're right Anders, we can see the raw log.  Forgot about that.  We're still getting a couple errors.  I'll look into those too, but we're getting closer.

**Comment by anders-dc** [comment]
Yes, we're getting there. Maybe it would be worth suppressing more compiler warnings. The main product of Travis is the boolean *yes/no* to whether the compilation and runtime tests for a commit are successful. Only rarely will somebody look into the log of a passed build.

**Comment by apcraig** [comment]
@anders-dc OK, there is another set of commits on the travis branch,
https://github.com/apcraig/CICE/tree/travis
specifically
https://github.com/apcraig/CICE/commit/6be69e45b7dbac1a104993bed6224b945db8368a

You should also add

setenv ICE_MACHINE_QUIETMODE  true

to your env.travisCI_gnu file.  that will stop the spewing of the build output.  if the build fails, it will do a tail -10 automatically on the build log file, so hopefully that will work for us.  if not, we'll continue to tweak.

In addition to adding the quiet mode, I have also added a couple tests to the travis suite.  I want to see what we get.  I have not been able to duplicate the error on another machine.  I even used the travisCI Macros file just to make sure it wasn't a small diff in the build settings.  I am getting some errors with other compilers (pgi) in what seems to be the same point, but I can't be sure it's the same thing.  I spent a few minutes looking at the pgi error but it's going to take a little more work to sort out.  My plan is to add an issue.

What I propose is we run this next set of tests and see what we get.  Then we should turn off, for now, the ones that are failing on travisCI.  We can then push this to master and separately work on the outstanding issues.  I think we've made some reasonable progress at this point.

**Comment by apcraig** [comment]
So, the latest test suite does more or less what I expected.  We definitely have some reproducibility problems, and that's one issue we're not seeing on other platforms so far.  That's even without OpenMP.  There is work to do, but most of that needs to happen outside Travis.  I propose the following changes to the travis_suite,  change

smoke          gx3     2x1        diag1,run5day   smoke_gx3_1x1_diag1_run5day
smoke          gx3     1x2        diag1,run5day
smoke          gx3     1x1        diag1,run5day,thread   smoke_gx3_1x2_diag1_run5day
smoke          gx3     2x1        diag1,run5day,thread   smoke_gx3_1x2_diag1_run5day

to

#smoke          gx3     2x1        diag1,run5day   smoke_gx3_1x1_diag1_run5day
smoke          gx3     2x1        diag1,run5day 
smoke          gx3     1x2        diag1,run5day
#smoke          gx3     1x1        diag1,run5day,thread   smoke_gx3_1x2_diag1_run5day
#smoke          gx3     2x1        diag1,run5day,thread   smoke_gx3_1x2_diag1_run5day
smoke          gx3     2x1        diag1,run5day,thread

Basically, we're turning off the 1x1 test that fails and turning off all the bfb compares for the other tests.  Not ideal, but OK for now.  @anders-dc can you make that change and retest.  If you prefer for me to make the change on my branch, just let me know.  thanks!

**Comment by anders-dc** [comment]
Hooray! [The build](https://travis-ci.org/anders-dc/CICE) took quite a long time to complete (33 mins), but passed. Thanks @apcraig!

**Comment by apcraig** [comment]
Great.  I think we an execute the PR now.  We should have @eclare108213 give a quick review too.  There are some code mods.  I may further reduce the test list or try to figure out a way for it to go a little faster.  30 minutes seems a little long for a "quick" status test.
